src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -0,0 +1,263 @@
+
+        private static OpenSearchKNNModelDao INSTANCE;
+        private static Client client;
+        private static ClusterService clusterService;
+        private static Settings settings;
+
+        /**
+         * Make sure we just have one instance of model index
+         *
+         * @return ModelIndex instance
+         */
+        public static synchronized OpenSearchKNNModelDao getInstance() {
+            if (INSTANCE == null) {
+                INSTANCE = new OpenSearchKNNModelDao();
+            }
+            return INSTANCE;
+        }
+
+        public static void initialize(Client client, ClusterService clusterService, Settings settings) {
+            OpenSearchKNNModelDao.client = client;
+            OpenSearchKNNModelDao.clusterService = clusterService;
+            OpenSearchKNNModelDao.settings = settings;
+        }
+
+        private OpenSearchKNNModelDao() {
+            numberOfShards = MODEL_INDEX_NUMBER_OF_SHARDS_SETTING.get(settings);
+            numberOfReplicas = MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING.get(settings);
+
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_INDEX_NUMBER_OF_SHARDS_SETTING,
+                    it -> numberOfShards = it);
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(MODEL_INDEX_NUMBER_OF_REPLICAS_SETTING,
+                    it -> numberOfReplicas = it);
+        }
+
+        @Override
+        public void create(ActionListener<CreateIndexResponse> actionListener) throws IOException {
+            if (isCreated()) {
+                return;
+            }
+
+            CreateIndexRequest request = new CreateIndexRequest(MODEL_INDEX_NAME)
+                    .mapping("_doc", getMapping(), XContentType.JSON)
+                    .settings(Settings.builder()
+                            .put("index.hidden", true)
+                            .put("index.number_of_shards", this.numberOfShards)
+                            .put("index.number_of_replicas", this.numberOfReplicas)
+                    );
+            client.admin().indices().create(request, actionListener);
+        }
+
+        @Override
+        public boolean isCreated() {
+            return clusterService.state().getRoutingTable().hasIndex(MODEL_INDEX_NAME);
+        }
+
+        @Override
+        public void put(String modelId, KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
+            String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+
+            Map<String, Object> parameters = ImmutableMap.of(
+                    KNNConstants.KNN_ENGINE, knnEngine.getName(),
+                    KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+            );
+
+            IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+            indexRequestBuilder.setId(modelId);
+            indexRequestBuilder.setSource(parameters);
+
+            put(indexRequestBuilder, listener);
```

Comment:
may be we can remove this private function and add it part of the same function. Better to keep all the IndexRequestBuilder parameters together to track easily and for better readability.

Reply:
Right, I was thinking about this. I was trying to remove code redundancy between the 2 put functions, but I think this leads to worse readability.